Pylint plugin to check return statements

KapJI opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
If function or method is annotated **explicit** return type `None`, `return` statements in it should be just `return` instead of `return None` or `return value`.

For example on:
```python3
def func():
    return None

def func2() -> None:
    return None

def func3() -> None:
    return

def func4() -> int | None:
    return None
```

This plugin will return:
> test.py:5:4: W0016: 'func2' is annotated with return type 'None' and returns value. Consider changing function return type or removing return value. (hass-return-none)

NOTE: In this example it only triggered on `func2` which has `-> None` but not on `func` or `func4`.

This can catch some cases when function is either having wrong return type annotation or some value is being returned without a need which in my opinion reduces code readability because such functions are meant to return nothing.

Inspired by https://github.com/home-assistant/core/pull/56390

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

KapJI committed:
Pylint plugin to validate empty returns

KapJI committed:
Support nested and async functions

KapJI committed:
Fix existing occurences

janiversen commented on homeassistant/util/__init__.py:
```diff
@@ -218,13 +218,13 @@ def __call__(self, method: Callable) -> Callable:
             async def throttled_value() -> None:
                 """Stand-in function for when real func is being throttled."""
-                return None
+                return None  # pylint: disable=hass-return-none
```
Why disable it here, you should be able to remove the None?

KapJI replied:
I think this is a very special case when function was created to return only `None` value. Although it can be removed the idea of that function that it returns `None` instead of no return value.

janiversen reviewed (approved):
Tradfri looks good.

cdce8p reviewed (commented):
The plugin does seem to work, but tbh I wonder if it's truly necessary. Checks aren't free, they always require additional time. One that visits every `return` even more so.

If `return None` would be an error, I would think of it differently. However, it's completely valid and just a matter of personal style.

balloob commented:
This feels weird. `return None` is more explicit and easier to read/understand.

KapJI commented:
To be honest I'm a bit surprised by received push back.

At least for me the function which has return type annotated with `-> None` means that it returns nothing, like `void` in other languages. Therefore it shouldn't return anything at all, at least explicitly, and callers shouldn't use the return value from these functions. And `return` in these functions should be only used to break execution earlier.

For example to me this:

```python3
def f(v):
    if v == 0:
        return
    print(v)
````

looks much better than explicit `None`:
```python3
def f(v):
    if v == 0:
        return None
    print(v)
````

I don't get why you prefer the second option.


This is different from functions which return optional where return value is expected and one should use explicit `return None`.

frenck commented:
> looks much better than explicit None:

I'm sorry, I don't agree. And I'm pretty sure there are more people in this project feeling the same way about it.

KapJI commented:
> And I'm pretty sure there are more people in this project feeling the same way about it.

Well, I think if it would then there would be much more places which were found by this plugin. In the last commit I fixed just 9 places with explicit `None` was used versus more than 3k places where `return` is used without the value. So it doesn't seem to be a common practice at all.

frenck commented:
> Well, I think if it would then there would be much more places which were found by this plugin.

Sure, but you don't need to convince me. I'm just leaving a comment that I don't like it to be implicit. 🤷‍♂️

janiversen commented:
I use both version (as KapJI knows) depending on the situation and I feel it is wrong to enforce it.

mypy is already slowing the commit quite a lot, and I would it was made faster not slower.

KapJI commented:
Well, this is pylint plugin and we don't run pylint during pre-commit hook. So this doesn't affect commit time at all.

On CI pylint job is not on a critical path either, so I don't think performance is a concern.

balloob commented on homeassistant/components/fritz/device_tracker.py:
```diff
@@ -60,8 +60,6 @@ async def async_get_scanner(hass: HomeAssistant, config: ConfigType) -> None:
         "Loading Fritz via scanner setup is now deprecated"
     )
 
-    return None
```
The problem I see is in cases like this, a `get_scanner` method that now returns nothing. Or a function that can return `X | None`, we should be able to just do `return None`. Relying on a pylint disable rule in such cases doesn't seem like a step forward.

If a return value of a function is never checked by a caller, we can omit `return None`, but I don't think that it's possible to detect / very common that it warrants a rule.

KapJI replied:
I think `get_scanner` doesn't have a good name here, it does indeed return nothing in any case:

````python3
async def async_get_scanner(hass: HomeAssistant, config: ConfigType) -> None:
```

This plugin only is only intended to catch cases when function is annotated with `-> None`. If function returns `X | None` then yes, `return None` should be used, this plugin won't complain about it.

balloob replied:
`async_get_scanner` should be annotated as `DeviceScanner | None` actually.

balloob replied:
In that case it is catching some cases, however it's catching mainly cases where the return type was bad – not the return value. And the way the warning is specified, the user will be inclined to remove the return statement and not fix the return type (as happened in this case here).

KapJI replied:
Right, the fix is not correct here, return type should be changed instead.

I agree, warning message should be corrected to include such case.

balloob replied:
Alright, in that case I agree it's an improvement, but we should make sure our warning message is clear that the issue can be with the return type OR the return value on the line that the issue was flagged.

MartinHjelmare replied:
@KapJI please clarify the PR description with your comment and example about optional return values. The current description doesn't explain this which is misleading.

KapJI commented on pylint/plugins/hass_return.py:
```diff
@@ -0,0 +1,45 @@
+    }
+    options = ()
+
+    def visit_return(self, node: Return) -> None:
```
Because I'm not that experienced with pylint/astroid to find undocumented `_get_return_nodes_skip_functions` :)
Thanks for the pointer!

I think pylint in many cases in used to force a good style rather than finding errors, so I don't see how this is different.

Regarding performance concerns I think it should be quite small and unnoticeable in a total run time. But I can try to measure the impact of extra plugin. Do you known what's the best way to see how long each individual check did take? Other than analyzing calls with cProfile.

cdce8p replied:
> Regarding performance concerns I think it should be quite small and unnoticeable in a total run time. But I can try to measure the impact of extra plugin. Do you known what's the best way to see how long each individual check did take? Other than analyzing calls with cProfile.

Performance profiling is a bit tricky, especially since at least some caching and multiprocessing is involved. I haven't tried it, but in theory you could write a decorator that sums the time all function calls to `visit_xxx` took. Just remember  to set `jobs=1` in `pyproject.toml` (for testing). I.e. something like this:
```py
from datetime import datetime, timedelta
from time import sleep
from typing import Callable

from typing_extensions import ParamSpec

P = ParamSpec("P")
runtime = timedelta(0)

def measure_time(func: Callable[P, None]) -> Callable[P, None]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> None:
        start = datetime.now()
        func(*args, **kwargs)
        global runtime
        runtime += datetime.now() - start
    return wrapper


@measure_time
def visit_xxx():
    sleep(3)


if __name__ == "__main__":
    print(runtime)
    visit_xxx()
    print(runtime)
```

I haven't actually tested that for a pylint `visit_xxx` method yet, but it should work. You can print the result in the `HassReturnFormatChecker.close()` method.

KapJI replied:
I did some tests and `_get_return_nodes_skip_functions` is about 60% slower than bottom-up implementation. This actually makes sense because `_get_return_nodes_skip_functions` iterates over all nodes in function subtree rather than just single path from return node to function root.

The plugin runtime is still very small, 120ms out of 10 minutes on my machine, which is ~0.02% of total runtime.

KapJI committed:
Fix device scanners return type

KapJI committed:
Update warning message

KapJI committed:
Use astroid.nodes

KapJI committed:
More fixes

KapJI commented:
I updated the description a bit and included example with optional. Hope it's more clear now.

Also I changed the logic to show the warning if function with `-> None` returns some value.

KapJI committed:
Use raise from

KapJI committed:
Include function name in the message

MartinHjelmare commented on homeassistant/components/zha/core/channels/security.py:
```diff
@@ -147,23 +147,23 @@ def _disarm(self, code: str):
         return zigbee_reply
 
-    def _arm_day(self, code: str) -> None:
+    def _arm_day(self, code: str):
```
We should correct the typing here and not remove it.

KapJI replied:
`zha` is excluded from type checking because it has many wrong type hints like this one: https://github.com/home-assistant/core/blob/dev/script/hassfest/mypy_config.py#L140

Correcting them is a large task which requires some domain understanding. I think absent type hints are better than incorrect ones.

MartinHjelmare replied:
CC @Adminiuga Can you give a change suggestion here?

Adminiuga replied:
Would this work? it really returns a coroutine 
```suggestion
    def _arm_day(self, code: str) -> awaitable:
```

MartinHjelmare commented on pylint/plugins/hass_constructor.py:
```diff
@@ -22,7 +22,7 @@ class HassConstructorFormatChecker(BaseChecker):  # type: ignore[misc]
     options = ()
 
-    def visit_functiondef(self, node: FunctionDef) -> None:
+    def visit_functiondef(self, node: nodes.FunctionDef) -> None:
```
Why was this change needed?

KapJI replied:
Based on the feedback in https://github.com/home-assistant/core/pull/56487#discussion_r713233925 I've decided to change it in all pylint plugins to have the same style everywhere: https://github.com/home-assistant/core/pull/56487/commits/9f9ad27a7146e45246f5bc6ac27b7bf195d1a475

cdce8p reviewed (commented):
Tbh I'm not really convinced this check will help. For the changes in this PR it might have identified one or two (potential ?) issues while as having some side effects. I highlighted some below.

Is the check really worth it?

MartinHjelmare commented on homeassistant/components/slack/notify.py:
```diff
@@ -308,7 +308,7 @@ async def async_send_message(self, message: str, **kwargs: Any) -> None:
                 blocks = None
 
-            return await self._async_send_text_only_message(
+            await self._async_send_text_only_message(
```
That's still possible with the check, but on a separate line, which is the correct way of doing it when not reruning an actual value.

MartinHjelmare commented on homeassistant/components/samsungtv/bridge.py:
```diff
@@ -197,7 +197,7 @@ def __init__(self, method: str, host: str, port: int | None) -> None:
         }
 
-    def mac_from_device(self) -> None:
+    def mac_from_device(self) -> str | None:
         """Try to fetch the mac address of the TV."""
         return None
```
This is an abstract method and the type annotation should match the parent method. The child can choose to return any of the types that the parent has specified. It doesn't need to return all types to stay within spec.

MartinHjelmare replied:
Ok, if the linters understand that, I agree it's better to use a more specific type annotation in the child while still staying within spec.

KapJI replied:
I think in such case you don't have to handle all possible types. If you know that method only returns `None` then you can just not use the returned value at all.

I agree, there could be a case when you don't know that for sure and rely on IDE suggestions but I find such cases extremely rare and there this check can be ignored. In fact in the whole codebase I haven't found any place where it could be useful.

cdce8p replied:
Let's agree to disagree on this one

KapJI committed:
Use Awaitable in zha
